extensions/amp-animation/0.1/parsers/css-expr-ast.js

```diff
@@ -1096,6 +1271,100 @@ export class CssCalcProductNode extends CssNode {
   }
 }
 
+/**
+ * CSS `min()` and `max()`.
+ * See https://developer.mozilla.org/en-US/docs/Web/CSS/min
+ * See https://developer.mozilla.org/en-US/docs/Web/CSS/max
+ */
+export class CssMinMaxNode extends CssFuncNode {
+  /**
+   * @param {string} name
+   * @param {!Array<!CssNode>} args
+   */
+  constructor(name, args) {
+    super(name, args);
+  }
+
+  /** @override */
+  isConst() {
+    return false;
+  }
+
+  /** @override */
+  calc(context, normalize) {
+    let resolvedArgs = resolveArray(context, normalize, this.args_, null);
+    if (!resolvedArgs) {
+      return null;
+    }
+
+    let firstNonPercent = null;
+    let hasPercent = false;
+    let hasDifferentUnits = false;
+    resolvedArgs.forEach((arg) => {
+      if (!(arg instanceof CssNumericNode)) {
+        throw new Error('arguments must be numerical');
+      }
+      if (arg instanceof CssPercentNode) {
+        hasPercent = true;
+      } else if (firstNonPercent) {
+        if (arg.type_ != firstNonPercent.type_) {
+          throw new Error('arguments must be the same type');
+        }
+        if (arg.units_ != firstNonPercent.units_) {
+          hasDifferentUnits = true;
+        }
+      } else {
+        firstNonPercent = arg;
+      }
+    });
+    if (firstNonPercent && hasPercent) {
+      hasDifferentUnits = true;
+    }
+
+    if (firstNonPercent) {
+      // Recalculate percent values and normalize units.
+      if (hasDifferentUnits) {
+        firstNonPercent = firstNonPercent.norm(context);
+      }
+      resolvedArgs = resolvedArgs.map((arg) => {
+        if (arg == firstNonPercent) {
+          return arg;
+        }
+
+        // Percent values.
+        if (arg instanceof CssPercentNode) {
+          return firstNonPercent.calcPercent(arg.num_, context);
+        }
+
+        // Units are the same, the math is simple: numerals are summed.
+        // Otherwise, the units neeed to be normalized first.
+        if (hasDifferentUnits) {
+          return /** @type {!CssNumericNode} */ (arg).norm(context);
+        }
+        return arg;
+      });
+    }
+
+    // Calculate.
+    const nums = resolvedArgs.map((arg) => arg.num_);
+    let value;
+    if (this.name_ == 'min') {
+      // min(...)
+      value = Math.min.apply(null, nums);
+    } else if (this.name_ == 'max') {
+      // max(...)
+      value = Math.max.apply(null, nums);
+    } else {
+      // clamp(min, preferred, max)
+      const min = nums[0];
+      const preferred = nums[1];
+      const max = nums[2];
+      value = Math.max(min, Math.min(max, preferred));
```

Comment:
Unfortunately, the CSS clamp is slightly different. Specifically, the wrong bounds are allowed, whereas `utils/math/clamp` throws an exception. And also because of this, the order of applying min and max is important.